### PR TITLE
Add Loggin infra to SlateDB

### LINKED ISSUE #90: Add some logging infra to SlateDB
We currently just log things by calling print!. We should instead convert print messages to calls to rust's logging facade (https://docs.rs/log/latest/log/) using the appropriate log level. Finally, we should set up some default logging implementation, like logging to stderr.

### PR DESCRIPTION
**Pull request by dharanad**
Closes #90

### CODE REVIEW

**Comment by dharanad** [comment]
@criccomini I just trying to see if i can improve this & add more relavant logs.

**Comment by criccomini** [comment]
K, LMK when you're ready for merge. :) Also, don't stress too much about getting the verbosity and log strings exactly right; that's a process that we can improve over time.

File: src/compaction_execute_bench.rs [hunk 1]
```diff
@@ -5,7 +5,7 @@ use crate::error::SlateDBError;
 use crate::sst::SsTableFormat;
 use crate::tablestore::TableStore;
 use crate::test_utils::OrderedBytesGenerator;
-use crate::{compactor::WorkerToOrchestoratorMsg, config::CompressionCodec};
+use crate::{compactor::WorkerToOrchestratorMsg, config::CompressionCodec};
```
Comment (rodesai) [inline src/compaction_execute_bench.rs:8]: thanks for fixing the typo!

**Comment by rodesai** [comment]
This looks great @dharanad ! Do you think you could add an example of how to set up the subscriber for recording the generated messages to the example in README.md? I also opened an issue (https://github.com/slatedb/slatedb/issues/129) to document the fact that we're using `tracing` in our docs.

File: Cargo.toml [hunk 2]
```diff
@@ -28,6 +28,7 @@ rand = "0.8.5"
 rand_xorshift = "0.3.0"
 log = {  version = "0.4.22", features = ["std"] }
 once_cell = "1.19.0"
+tracing = "0.1.40"
```
Comment (rodesai) [inline Cargo.toml:31]: do we need to enable the "log" feature so that log records are generated?

Comment (wasup-yash) [reply Cargo.toml:31]: are you talking about the general log that we see or specific time period based logs cause then we can work with span and events (part of tracing crate).

Comment (rodesai) [reply Cargo.toml:31]: from https://docs.rs/tracing/latest/tracing/#emitting-log-records:

```
This crate provides two feature flags, “log” and “log-always”, which will cause [spans](https://docs.rs/tracing/latest/tracing/span/index.html) and [events](https://docs.rs/tracing/latest/tracing/event/struct.Event.html) to emit log records. When the “log” feature is enabled, if no tracing Subscriber is active, invoking an event macro or creating a span with fields will emit a log record. This is intended primarily for use in libraries which wish to emit diagnostics that can be consumed by applications using tracing or log, without paying the additional overhead of emitting both forms of diagnostics when tracing is in use.
```

It sounds like the `log` feature needs to be enabled to actually emit log records (as emitted by the basic rust logging facade (https://docs.rs/log/latest/log/)) when using `tracing`?

Comment (wasup-yash) [reply Cargo.toml:31]: yes there is a flag feature as you mentioned but also there is [this](https://docs.rs/tracing/latest/tracing/#crate-feature-flags:~:text=tracing%2Dsubscriber%20provides%20Subscriber%20implementations%20and%20utilities%20for%20working%20with%20Subscribers.%20This%20includes%20a%20FmtSubscriber%20FmtSubscriber%20for%20logging%20formatted%20trace%20data%20to%20stdout%2C%20with%20similar%20filtering%20and%20formatting%20to%20the%20env_logger%20crate) which we would want , also enabling feature of log we will have logs in text base format with field from level TRACE not from the actual level it is happening from you can see [this](https://docs.rs/tracing/latest/tracing/#emitting-log-records:~:text=.%20Since%20these%20additional%20span%20lifecycle%20logs%20have%20the%20potential%20to%20be%20very%20verbose%2C%20and%20don%E2%80%99t%20include%20additional%20fields%2C%20they%20will%20always%20be%20emitted%20at%20the%20Trace%20level%2C%20rather%20than%20inheriting%20the%20level%20of%20the%20span%20that%20generated%20them.)

Comment (rodesai) [reply Cargo.toml:31]: > yes there is a flag feature as you mentioned but also there is [this](https://docs.rs/tracing/latest/tracing/#crate-feature-flags:~:text=tracing%2Dsubscriber%20provides%20Subscriber%20implementations%20and%20utilities%20for%20working%20with%20Subscribers.%20This%20includes%20a%20FmtSubscriber%20FmtSubscriber%20for%20logging%20formatted%20trace%20data%20to%20stdout%2C%20with%20similar%20filtering%20and%20formatting%20to%20the%20env_logger%20crate) which we would want

Not sure what the suggestion is here. I agree you'd need to set up a subscriber to actually consume and record the logs somewhere useful. But I think that's up to the user of slatedb to import and set up. Since SlateDB is a library, it probably shouldn't set up any subscriber itself.

> also enabling feature of log we will have logs in text base format with field from level TRACE not from the actual level it is happening from you can see [this](https://docs.rs/tracing/latest/tracing/#emitting-log-records:~:text=.%20Since%20these%20additional%20span%20lifecycle%20logs%20have%20the%20potential%20to%20be%20very%20verbose%2C%20and%20don%E2%80%99t%20include%20additional%20fields%2C%20they%20will%20always%20be%20emitted%20at%20the%20Trace%20level%2C%20rather%20than%20inheriting%20the%20level%20of%20the%20span%20that%20generated%20them.)

I think all that's saying is that when the log feature is enabled, `tracing` will create log records at TRACE whenever a span is entered/exited/closed. Spans are different from events, and denote when certain contexts are entered/exited. For example, you could create a span that's entered whenever a function F is called and exited whenever the function returns. Within that span you could emit multiple events (which are the result of the calls to info!, error!, etc). It's likely that spans are generally much more frequently entered/exited than logs are emitted. E.G you may call F 1000 times but only every log once when some error happens. So that's saying the span enter/exit message will always be at TRACE level so you don't log too much.

Comment (wasup-yash) [reply Cargo.toml:31]: yes thanks for the explanation I was not suggesting but trying to understand what we would want to include and what not `Since SlateDB is a library, it probably shouldn't set up any subscriber itself.` this is the crux, we would just need minimal log feature then which happens occasionally ;)

Comment (criccomini) [reply Cargo.toml:31]: Getting up to speed on `tracing`. My first time using it. I agree with @rodesai. The way I read the [Emitting `log` Records](https://docs.rs/tracing/latest/tracing/#emitting-log-records) section, it sounds like we want either `log` or `log-always`:

> This is intended primarily for use in libraries which wish to emit diagnostics that can be consumed by applications using tracing or log, without paying the additional overhead of emitting both forms of diagnostics when tracing is in use.

I think this is what we want. If no consumer is set up, just default to `log`. I don't think we want `log-always`, since that would cause both `log` and `tracing` events to be emitted:

> Unlike the “log” feature, libraries generally should not enable the “log-always” feature, as doing so will prevent applications from being able to opt out of the log records.

tl;dr I think adding the `[log]` feature in `Cargo.toml` is the way to go.

**Comment by vigneshc** [comment]
> This looks great @dharanad ! Do you think you could add an example of how to set up a subscriber for recording the generated messages to the example in README.md? I also opened an issue (#129) to document the fact that we're using `tracing` in our docs.

**Comment by vigneshc** [comment]
Closed it accidentally.

**Comment by vigneshc** [comment]
Changes look good.
+1 on @rodesai 's comment earlier on adding example in Readme to subscribe to the logs.
Also, would failing tests automatically emit these traces or would we need to pass in additional parameters?

File: src/compaction_execute_bench.rs [hunk 3]
```diff
@@ -296,7 +297,7 @@ fn load_options() -> Options {
         val_bytes: val_bytes.parse::<usize>().expect("invalid val bytes"),
         compression_codec,
     };
-    println!("Options: {:#?}", options);
+    info!("Options: {:#?}", options);
```
Comment (vigneshc) [inline src/compaction_execute_bench.rs:300]: Should we continue using formatted strings or should we use structured events?
This is first time I am looking tracing crate, trying to learn what would be better.

Can you share examples of how these traces look like in the console subscriber?

**Comment by rodesai** [comment]
> Also, would failing tests automatically emit these traces or would we need to pass in additional parameters?

Ooh this is a great point. We'd probably need to make sure that there is some subscriber set up that records messages on stderr/stdout for tests.

Also @dharanad any reason to keep this as a draft? I think once the concerns around tests and the `log` feature are addressed (which the test feedback probably will require anyway) we should be good to merge this.

File: Cargo.toml (hunk 2 above)
Comment (dharanad) [reply Cargo.toml:31]: +1 to enabling `log` feature. The `log` feature in the tracing crate allows tracing to interoperate with the log crate, which is the traditional logging system in Rust.

File: src/compaction_execute_bench.rs (hunk 3 above)
Comment (dharanad) [reply src/compaction_execute_bench.rs:300]:
```
#[derive(Debug)]
struct Baz {
    string: String,
    int: i32,
    bool_value: bool,
    now: Instant
}
```

```
info!("Baz value {:?}", b);
```


```
2024-08-18T04:25:18.241655Z  INFO tracing_demo: Baz value Baz {
    string: "slate",
    int: 123,
    bool_value: false,
    now: Instant {
        tv_sec: 361652,
        tv_nsec: 925992000,
    },
}
```

**Comment by dharanad** [comment]
> Changes look good. +1 on @rodesai 's comment earlier on adding example in Readme to subscribe to the logs. Also, would failing tests automatically emit these traces or would we need to pass in additional parameters?

I think we can make use of [tracing_test](https://docs.rs/tracing-test/latest/tracing_test/) & annotate the test with `#[traced_test]` for now.

**Comment by dharanad** [comment]
Apologies for the delay. I was unwell.

File: src/compaction_execute_bench.rs (hunk 3 above)
Comment (criccomini) [reply src/compaction_execute_bench.rs:300]: Not critical to this conversation, but multi-line log lines are a bummer. Is it possible to configure `log` to emit single line?

**Comment by criccomini** [comment]
No worries at all. Hope you're feeling better. :)

Looks like you might need another rebase. Seeing some test failures related to the compression stuff we merged in. Main is passing now.

Regarding `tracing_test`, that looks more like it's meant for testing the actual log lines:

> The focus is on testing the logging, not on debugging the tests

It contains methods like `logs_contain`.

I think we might want either [env_logger](https://docs.rs/env_logger/latest/env_logger/#capturing-logs-in-tests) or [test-log](https://crates.io/crates/test-log).

File: src/compaction_execute_bench.rs (hunk 3 above)
Comment (dharanad) [reply src/compaction_execute_bench.rs]: I have remove pretty print formatting, this should emit log in single line

**Comment by rodesai** [comment]
> I think we might want either [env_logger](https://docs.rs/env_logger/latest/env_logger/#capturing-logs-in-tests) or [test-log](https://crates.io/crates/test-log).

I've been using `test-log` for another project and it seems like a reasonable solution. It's a bit annoying because you have t use their test macro for every test, but I haven't seen a better approach.

**Comment by rodesai** [comment]
@dharanad mind if we merge this? I could use the changes here.

**Comment by dharanad** [comment]
> @dharanad mind if we merge this? I could use the changes here.

No. Please go ahead

**Comment by rodesai** [comment]
@dharanad you need to fix the `Cargo.toml` file. You probably just need to rebase onto main.

**Comment by criccomini** [comment]
I've re-added the `Cargo.toml` deletions back. Looks like we're all good.
